crates/fuel-core/src/service/adapters/consensus_module/poa/pre_confirmation_signature/trigger.rs

```diff
@@ -0,0 +1,135 @@
+use fuel_core_poa::{
+    ports::GetTime,
+    pre_confirmation_signature_service::{
+        error::Result as PoAResult,
+        trigger::KeyRotationTrigger,
+    },
+};
+use fuel_core_types::tai64::Tai64;
+use std::time::Duration;
+
+pub struct TimeBasedTrigger<Time> {
+    time: Time,
+    next_rotation: Tai64,
+    rotation_interval: Duration,
+}
+
+impl<Time: GetTime> TimeBasedTrigger<Time> {
+    pub fn new(time: Time, rotation_interval: Duration) -> Self {
+        let now = time.now();
+        let _next_rotation = now.0.saturating_add(rotation_interval.as_secs());
+        let next_rotation = Tai64(_next_rotation);
+        Self {
+            time,
+            next_rotation,
+            rotation_interval,
+        }
+    }
+
+    fn get_next_rotation(&self) -> PoAResult<Tai64> {
+        Ok(self.next_rotation)
+    }
+
+    fn set_next_rotation(&mut self, next_rotation: Tai64) -> PoAResult<()> {
+        self.next_rotation = next_rotation;
+        Ok(())
+    }
+}
+
+fn duration_between(t1: Tai64, t2: Tai64) -> Duration {
+    let diff = t2.0.saturating_sub(t1.0);
+    Duration::from_secs(diff)
+}
+
+impl<Time: GetTime + Send> KeyRotationTrigger for TimeBasedTrigger<Time> {
+    async fn next_rotation(&mut self) -> PoAResult<()> {
+        // create a future that will resolve at the next rotation time
+        // if that future resolves, update the next rotation time
+        let next_rotation = self.get_next_rotation()?;
+        let time_to_wait = duration_between(self.time.now(), next_rotation);
+        tokio::time::sleep(time_to_wait).await;
```

Comment:
If we do that, does it means that the service `pre_confirmation_signature` is stuck before we reach the next signing `Tai64`, which prevents us from signing pre-confirmation or it's fine ? Because I believe both are on the same `tokio::task` and so we would be stuck ?

Reply:
I think we should just use timer and `tick().await` here=)